src/Cli/dotnet/commands/dotnet-workload/install/WorkloadManifestUpdater.cs

```diff
@@ -0,0 +1,172 @@
+            }
+        }
+
+        public IEnumerable<(ManifestId manifestId, ManifestVersion existingVersion, ManifestVersion newVersion)> CalculateManifestUpdates(SdkFeatureBand featureBand)
+        {
+            var manifestUpdates = new List<(ManifestId, ManifestVersion, ManifestVersion)>();
+            var currentManifestIds = GetInstalledManifestIds();
+            foreach (var manifestId in currentManifestIds)
+            {
+                var currentManifestVersion = GetInstalledManifestVersion(manifestId);
+                var adManifestVersion = GetAdvertisingManifestVersion(featureBand, manifestId);
+                if (adManifestVersion == null)
+                {
+                    continue;
+                }
+
+                if (adManifestVersion != null && adManifestVersion.CompareTo(currentManifestVersion) > 0)
+                {
+                    manifestUpdates.Add((manifestId, currentManifestVersion, adManifestVersion));
+                }
+            }
+            return manifestUpdates;
+        }
+
+        private IEnumerable<ManifestId> GetInstalledManifestIds()
+        {
+            var manifestDirs = _workloadManifestProvider.GetManifestDirectories();
+
+            var manifests = new List<ManifestId>();
+            foreach (var manifestDir in manifestDirs)
+            {
+                var manifestId = Path.GetFileName(manifestDir);
+                manifests.Add(new ManifestId(manifestId));
+            }
+            return manifests;
+        }
+
+        private async Task UpdateAdvertisingManifestAsync(ManifestId manifestId, SdkFeatureBand featureBand, bool includePreviews)
+        {
+            string packagePath = null;
+            string extractionPath = null;
+            try
+            {
+                var adManifestPath = GetAdvertisingManifestPath(featureBand, manifestId);
+                packagePath = await _nugetPackageDownloader.DownloadPackageAsync(GetManifestPackageId(featureBand, manifestId), includePreview: includePreviews);
+                extractionPath = Path.Combine(_userHome, ".dotnet", "sdk-advertising-temp", $"{manifestId}-extracted");
+                Directory.CreateDirectory(extractionPath);
+                var resultingFiles = await _nugetPackageDownloader.ExtractPackageAsync(packagePath, extractionPath);
+
+                if (Directory.Exists(adManifestPath))
+                {
+                    Directory.Delete(adManifestPath, true);
+                }
+                Directory.CreateDirectory(Path.GetDirectoryName(adManifestPath));
+                FileAccessRetrier.RetryOnMoveAccessFailure(() => Directory.Move(Path.Combine(extractionPath, "data"), adManifestPath));
+
+                _reporter.WriteLine(string.Format(LocalizableStrings.AdManifestUpdated, manifestId));
+            }
+            catch (Exception e)
+            {
+                _reporter.WriteLine(string.Format(LocalizableStrings.FailedAdManifestUpdate, manifestId, e.Message));
```

Comment:
It sounds like the failure we would expect is if there are no manifest packages available on the NuGet feed.  I think we should check for that explicitly and treat it not as an error (ie when we call DownloadPackageAsync).  I think other failures are probably not expected and we may want to fail.